Shell Absolute Navigation Crashes when a title view is set to each shell page.

### Description

If you add a title view to the pages inside a shell as I did in the reproduction sample, then you perform absolute navigation, you're going to face a crash. With the following stack trace errors, I pasted in the log output below.





### Steps to Reproduce

1- Create a new MAUI app
2- Add a shell and two pages inside a tabbar
3- Add a title view using Shell.SetTitleView
4- Perform absolute navigation from one page to the other, and the app will crash.

Here is a link to [the reproduction of this bug](https://github.com/DamienDoumer/MAUIBugs/tree/master/MAUIBugs/ShellNavBugOne).

### Version with bug

Preview 14 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android

### Did you find any workaround?

The work around is to not use Shell.SetTitleView, and code the title view directly in the page's view hierarchy

### Relevant log output

```shell
`[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: Java.Lang.IllegalStateException: The specified child already has a parent. You must call removeView() on the child's parent first.
[mono-rt]    at Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualVoidMethod(JniObjectReference instance, JniObjectReference type, JniMethodInfo method, JniArgumentValue* args) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniEnvironment.g.cs:line 12324
[mono-rt]    at Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod(String encodedMember, IJavaPeerable self, JniArgumentValue* parameters) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniInstanceMethods_Invoke.cs:line 35
[mono-rt]    at Android.Views.ViewGroup.AddView(View child) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net6.0/android-32/mcw/Android.Views.ViewGroup.cs:line 1999
[mono-rt]    at Microsoft.Maui.Handlers.LayoutHandler.SetVirtualView(IView view) in D:\a\_work\1\s\src\Core\src\Handlers\Layout\LayoutHandler.Android.cs:line 43
[mono-rt]    at Microsoft.Maui.Handlers.ViewHandler`2[[Microsoft.Maui.ILayout, Microsoft.Maui, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[Microsoft.Maui.Platform.LayoutViewGroup, Microsoft.Maui, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].SetVirtualView(IElement view) in D:\a\_work\1\s\src\Core\src\Handlers\View\ViewHandlerOfT.cs:line 54
[mono-rt]    at Microsoft.Maui.Controls.Element.SetHandler(IElementHandler newHandler) in D:\a\_work\1\s\src\Controls\src\Core\HandlerImpl\Element\Element.Impl.cs:line 64
[mono-rt]    at Microsoft.Maui.Controls.Element.set_Handler(IElementHandler value) in D:\a\_work\1\s\src\Controls\src\Core\HandlerImpl\Element\Element.Impl.cs:line 20
[mono-rt]    at Microsoft.Maui.Controls.VisualElement.Microsoft.Maui.IElement.set_Handler(IElementHandler value) in D:\a\_work\1\s\src\Controls\src\Core\HandlerImpl\VisualElement\VisualElement.Impl.cs:line 291
[mono-rt]    at Microsoft.Maui.Platform.ElementExtensions.ToHandler(IElement view, IMauiContext context) in D:\a\_work\1\s\src\Core\src\Platform\ElementExtensions.cs:line 91
[mono-rt]    at Microsoft.Maui.Platform.ElementExtensions.ToPlatform(IElement view, IMauiContext context) in D:\a\_work\1\s\src\Core\src\Platform\ElementExtensions.cs:line 127
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellViewRenderer.OnViewSet(View view) in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ShellContentView.cs:line 207
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellViewRenderer.set_View(View value) in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ShellContentView.cs:line 46
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellViewRenderer..ctor(Context context, View view, IMauiContext mauiContext) in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ShellContentView.cs:line 38
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ContainerView.OnViewSet(View view) in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ContainerView.cs:line 117
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ContainerView.set_View(View value) in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ContainerView.cs:line 38
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ContainerView..ctor(Context context, View view, IMauiContext mauiContext) in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ContainerView.cs:line 20
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellToolbarTracker.UpdateTitleView(Context context, Toolbar toolbar, View titleView) in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ShellToolbarTracker.cs:line 539
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellToolbarTracker.UpdateTitleView() in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ShellToolbarTracker.cs:line 656
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellToolbarTracker.OnPageChanged(Page oldPage, Page newPage) in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ShellToolbarTracker.cs:line 249
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellToolbarTracker.set_Page(Page value) in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ShellToolbarTracker.cs:line 105
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellSectionRenderer.OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ShellSectionRenderer.cs:line 174
[mono-rt]    at AndroidX.Fragment.App.Fragment.n_OnCreateView_Landroid_view_LayoutInflater_Landroid_view_ViewGroup_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_inflater, IntPtr native_container, IntPtr native_savedInstanceState) in D:\a\1\s\generated\androidx.fragment.fragment\obj\Release
et6.0-android\generated\src\AndroidX.Fragment.App.Fragment.cs:line 1977
[mono-rt]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPLLL_L(_JniMarshal_PPLLL_L callback, IntPtr jnienv, IntPtr klazz, IntPtr p0, IntPtr p1, IntPtr p2) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 231
[mono-rt]   --- End of managed Java.Lang.IllegalStateException stack trace ---
[mono-rt] java.lang.IllegalStateException: The specified child already has a parent. You must call removeView() on the child's parent first.
[mono-rt] 	at android.view.ViewGroup.addViewInner(ViewGroup.java:5319)
[mono-rt] 	at android.view.ViewGroup.addView(ViewGroup.java:5140)
[mono-rt] 	at android.view.ViewGroup.addView(ViewGroup.java:5080)
[mono-rt] 	at android.view.ViewGroup.addView(ViewGroup.java:5052)
[mono-rt] 	at crc648677f43942f779a4.ShellSectionRenderer.n_onCreateView(Native Method)
[mono-rt] 	at crc648677f43942f779a4.ShellSectionRenderer.onCreateView(ShellSectionRenderer.java:42)
[mono-rt] 	at androidx.fragment.app.Fragment.performCreateView(Fragment.java:2963)
[mono-rt] 	at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:518)
[mono-rt] 	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:282)
[mono-rt] 	at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2189)
[mono-rt] 	at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2100)
[mono-rt] 	at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:2002)
[mono-rt] 	at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:524)
[mono-rt] 	at android.os.Handler.handleCallback(Handler.java:938)
[mono-rt] 	at android.os.Handler.dispatchMessage(Handler.java:99)
[mono-rt] 	at android.os.Looper.loopOnce(Looper.java:233)
[mono-rt] 	at android.os.Looper.loop(Looper.java:344)
[mono-rt] 	at android.app.ActivityThread.main(ActivityThread.java:8184)
[mono-rt] 	at java.lang.reflect.Method.invoke(Native Method)
[mono-rt] 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:584)
[mono-rt] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1034)
[mono-rt] 
[mono-rt]   --- End of managed Java.Lang.IllegalStateException stack trace ---
[mono-rt] java.lang.IllegalStateException: The specified child already has a parent. You must call removeView() on the child's parent first.
[mono-rt] 	at android.view.ViewGroup.addViewInner(ViewGroup.java:5319)
[mono-rt] 	at android.view.ViewGroup.addView(ViewGroup.java:5140)
[mono-rt] 	at android.view.ViewGroup.addView(ViewGroup.java:5080)
[mono-rt] 	at android.view.ViewGroup.addView(ViewGroup.java:5052)
[mono-rt] 	at crc648677f43942f779a4.ShellSectionRenderer.n_onCreateView(Native Method)
[mono-rt] 	at crc648677f43942f779a4.ShellSectionRenderer.onCreateView(ShellSectionRenderer.java:42)
[mono-rt] 	at androidx.fragment.app.Fragment.performCreateView(Fragment.java:2963)
[mono-rt] 	at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:518)
[mono-rt] 	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:282)
[mono-rt] 	at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2189)
[mono-rt] 	at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2100)
[mono-rt] 	at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:2002)
[mono-rt] 	at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:524)
[mono-rt] 	at android.os.Handler.handleCallback(Handler.java:938)
[mono-rt] 	at android.os.Handler.dispatchMessage(Handler.java:99)
[mono-rt] 	at android.os.Looper.loopOnce(Looper.java:233)
[mono-rt] 	at android.os.Looper.loop(Looper.java:344)
[mono-rt] 	at android.app.ActivityThread.main(ActivityThread.java:8184)
[mono-rt] 	at java.lang.reflect.Method.invoke(Native Method)
[mono-rt] 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:584)
[mono-rt] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1034)
[mono-rt] 
`
```

--- a/src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellToolbarTracker.cs
+++ b/src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellToolbarTracker.cs
@@ -198,6 +198,8 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 				_drawerLayout.RemoveDrawerListener(_drawerToggle);
 				_drawerToggle?.Dispose();
 
+				_toolbar?.Handler?.DisconnectHandler();
+				_toolbar = null;
 				_platformToolbar.RemoveAllViews();
 			}
 

--- a/src/Controls/src/Core/HandlerImpl/Toolbar/Toolbar.Android.cs
+++ b/src/Controls/src/Core/HandlerImpl/Toolbar/Toolbar.Android.cs
@@ -9,13 +9,14 @@ using Android.Views;
 using Google.Android.Material.AppBar;
 using Microsoft.Maui.Controls.Platform;
 using Microsoft.Maui.Handlers;
+using LP = Android.Views.ViewGroup.LayoutParams;
 
 namespace Microsoft.Maui.Controls
 {
 	public partial class Toolbar
 	{
-		IViewHandler? _nativeTitleViewHandler;
-		Container? _nativeTitleView;
+		IViewHandler? _platformTitleViewHandler;
+		Container? _platformTitleView;
 		List<IMenuItem> _currentMenuItems = new List<IMenuItem>();
 		List<ToolbarItem> _currentToolbarItems = new List<ToolbarItem>();
 
@@ -23,6 +24,17 @@ namespace Microsoft.Maui.Controls
 			Handler?.MauiContext?.GetNavigationRootManager();
 
 		MaterialToolbar PlatformView => Handler?.PlatformView as MaterialToolbar ?? throw new InvalidOperationException("Native View not set");
+
+		partial void OnHandlerChanging(IElementHandler oldHandler, IElementHandler newHandler)
+		{
+			if (newHandler == null)
+			{
+				if (_platformTitleView != null)
+					_platformTitleView.Child = null;
+
+				_platformTitleViewHandler?.DisconnectHandler();
+			}
+		}
 
 		void OnToolbarItemPropertyChanged(object? sender, PropertyChangedEventArgs e)
 		{
@@ -57,39 +69,45 @@ namespace Microsoft.Maui.Controls
 			_ = MauiContext.Context ?? throw new ArgumentNullException(nameof(MauiContext.Context));
 
 			VisualElement titleView = TitleView;
-			if (_nativeTitleViewHandler != null)
+			if (_platformTitleViewHandler != null)
 			{
-				var reflectableType = _nativeTitleViewHandler as System.Reflection.IReflectableType;
-				var rendererType = reflectableType != null ? reflectableType.GetTypeInfo().AsType() : _nativeTitleViewHandler.GetType();
-				if (titleView == null || Internals.Registrar.Registered.GetHandlerTypeForObject(titleView) != rendererType)
+				Type? rendererType = null;
+
+				if (titleView != null)
+					rendererType = MauiContext.Handlers.GetHandlerType(titleView.GetType());
+
+				if (titleView == null || titleView.Handler?.GetType() != rendererType)
 				{
-					if (_nativeTitleView != null)
-						_nativeTitleView.Child = null;
+					if (_platformTitleView != null)
+						_platformTitleView.Child = null;
 
-					if (_nativeTitleViewHandler?.VirtualView != null)
-						_nativeTitleViewHandler.VirtualView.Handler = null;
+					if (_platformTitleViewHandler?.VirtualView != null)
+						_platformTitleViewHandler.VirtualView.Handler = null;
 
-					_nativeTitleViewHandler = null;
+					_platformTitleViewHandler = null;
 				}
 			}
 
 			if (titleView == null)
 				return;
 
-			if (_nativeTitleViewHandler != null)
-				_nativeTitleViewHandler.SetVirtualView(titleView);
+			if (_platformTitleViewHandler != null)
+				_platformTitleViewHandler.SetVirtualView(titleView);
 			else
 			{
 				titleView.ToPlatform(MauiContext);
-				_nativeTitleViewHandler = titleView.Handler;
+				_platformTitleViewHandler = titleView.Handler;
 
-				if (_nativeTitleView == null)
+				if (_platformTitleView == null)
 				{
-					_nativeTitleView = new Container(MauiContext.Context);
-					PlatformView.AddView(_nativeTitleView);
+					var context = MauiContext.Context!;
+					_platformTitleView = new Container(context);
+					var layoutParams = new MaterialToolbar.LayoutParams(LP.MatchParent, LP.MatchParent);
+					_platformTitleView.LayoutParameters = layoutParams;
+					PlatformView.AddView(_platformTitleView);
 				}
 
-				_nativeTitleView.Child = (IPlatformViewHandler?)_nativeTitleViewHandler;
+				_platformTitleView.Child = (IPlatformViewHandler?)_platformTitleViewHandler;
 			}
 		}
 
@@ -155,13 +173,18 @@ namespace Microsoft.Maui.Controls
 			{
 				set
 				{
-					if (_child != null)
-						RemoveView(_child.PlatformView);
+					_child?.DisconnectHandler();
+					RemoveAllViews();
 
 					_child = value;
 
-					if (value != null)
-						AddView(value.PlatformView);
+					if (_child != null)
+					{
+						var platformView = _child.ToPlatform();
+						platformView.RemoveFromParent();
+						if (platformView != null)
+							AddView(platformView);
+					}
 				}
 			}
 
@@ -170,7 +193,8 @@ namespace Microsoft.Maui.Controls
 				if (_child?.PlatformView == null)
 					return;
 
-				_child.PlatformView.Layout(l, t, r, b);
+				var destination = Context!.ToCrossPlatformRectInReferenceFrame(l, t, r, b);
+				_child?.VirtualView?.Arrange(destination);
 			}
 
 			protected override void OnMeasure(int widthMeasureSpec, int heightMeasureSpec)
@@ -181,8 +205,17 @@ namespace Microsoft.Maui.Controls
 					return;
 				}
 
-				_child.PlatformView.Measure(widthMeasureSpec, heightMeasureSpec);
-				SetMeasuredDimension(_child.PlatformView.MeasuredWidth, _child.PlatformView.MeasuredHeight);
+				var width = widthMeasureSpec.GetSize();
+				var height = heightMeasureSpec.GetSize();
+
+				if (width * height == 0)
+				{
+					SetMeasuredDimension(0, 0);
+					return;
+				}
+
+				_child?.VirtualView?.Measure(widthMeasureSpec.ToDouble(Context!), heightMeasureSpec.ToDouble(Context!));
+				SetMeasuredDimension(width, height);
 			}
 		}
 	}

--- a/src/Controls/src/Core/NavigationPageToolbar.cs
+++ b/src/Controls/src/Core/NavigationPageToolbar.cs
@@ -230,7 +230,8 @@ namespace Microsoft.Maui.Controls
 
 		Color GetBarTextColor() => _currentNavigationPage?.BarTextColor;
 		Color GetIconColor() => (_currentPage != null) ? NavigationPage.GetIconColor(_currentPage) : null;
-		string GetTitle() => _currentPage?.Title;
+		string GetTitle() => GetTitleView() != null ? String.Empty : _currentPage?.Title;
+
 		VisualElement GetTitleView()
 		{
 			if (_currentNavigationPage == null)

--- a/src/Controls/src/Core/Platform/Windows/Extensions/ToolbarExtensions.cs
+++ b/src/Controls/src/Core/Platform/Windows/Extensions/ToolbarExtensions.cs
@@ -11,71 +11,80 @@ namespace Microsoft.Maui.Controls.Platform
 {
 	internal static class ToolbarExtensions
 	{
-		public static void UpdateIsVisible(this MauiToolbar nativeToolbar, Toolbar toolbar)
+		public static void UpdateIsVisible(this MauiToolbar platformToolbar, Toolbar toolbar)
 		{
-			nativeToolbar.Visibility = (toolbar.IsVisible) ? UI.Xaml.Visibility.Visible : UI.Xaml.Visibility.Collapsed;
+			platformToolbar.Visibility = (toolbar.IsVisible) ? UI.Xaml.Visibility.Visible : UI.Xaml.Visibility.Collapsed;
 		}
 
-		public static void UpdateTitleIcon(this MauiToolbar nativeToolbar, Toolbar toolbar)
+		public static void UpdateTitleIcon(this MauiToolbar platformToolbar, Toolbar toolbar)
 		{
 			_ = toolbar?.Handler?.MauiContext ?? throw new ArgumentNullException(nameof(toolbar.Handler.MauiContext));
 			toolbar.TitleIcon.LoadImage(toolbar.Handler.MauiContext, (result) =>
 			{
 				if (result != null)
 				{
-					nativeToolbar.TitleIconImageSource = result.Value;
+					platformToolbar.TitleIconImageSource = result.Value;
 					toolbar.Handler.UpdateValue(nameof(Toolbar.IconColor));
 				}
 				else
-					nativeToolbar.TitleIconImageSource = null;
+					platformToolbar.TitleIconImageSource = null;
 			});
 		}
 
-		public static void UpdateBackButton(this MauiToolbar nativeToolbar, Toolbar toolbar)
+		public static void UpdateBackButton(this MauiToolbar platformToolbar, Toolbar toolbar)
 		{
-			nativeToolbar.IsBackEnabled =
+			platformToolbar.IsBackEnabled =
 				toolbar.BackButtonEnabled && toolbar.BackButtonVisible;
 
-			nativeToolbar
+			platformToolbar
 				.IsBackButtonVisible = (toolbar.BackButtonVisible) ? NavigationViewBackButtonVisible.Visible : NavigationViewBackButtonVisible.Collapsed;
 
 			toolbar.Handler?.UpdateValue(nameof(Toolbar.BarBackground));
 		}
 
-		public static void UpdateBarBackground(this MauiToolbar nativeToolbar, Toolbar toolbar)
+		public static void UpdateBarBackground(this MauiToolbar platformToolbar, Toolbar toolbar)
 		{
-			nativeToolbar.Background = toolbar.BarBackground?.ToBrush();
+			platformToolbar.Background = toolbar.BarBackground?.ToBrush();
 		}
 
-		public static void UpdateTitleView(this MauiToolbar nativeToolbar, Toolbar toolbar)
+		public static void UpdateTitleView(this MauiToolbar platformToolbar, Toolbar toolbar)
 		{
 			_ = toolbar.Handler?.MauiContext ?? throw new ArgumentNullException(nameof(toolbar.Handler.MauiContext));
 
-			nativeToolbar.TitleView = toolbar.TitleView?.ToPlatform(toolbar.Handler.MauiContext);
+			platformToolbar.TitleView = toolbar.TitleView?.ToPlatform(toolbar.Handler.MauiContext);
+
+			if (toolbar.TitleView is IView view)
+			{
+				platformToolbar.TitleViewMargin = view.Margin.ToPlatform();
+			}
+			else
+			{
+				platformToolbar.TitleViewMargin = new UI.Xaml.Thickness(0);
+			}
 		}
 
-		public static void UpdateIconColor(this MauiToolbar nativeToolbar, Toolbar toolbar)
+		public static void UpdateIconColor(this MauiToolbar platformToolbar, Toolbar toolbar)
 		{
 			// This property wasn't wired up in Controls
 		}
 
-		public static void UpdateTitle(this MauiToolbar nativeToolbar, Toolbar toolbar)
+		public static void UpdateTitle(this MauiToolbar platformToolbar, Toolbar toolbar)
 		{
-			nativeToolbar.Title = toolbar.Title;
+			platformToolbar.Title = toolbar.Title;
 		}
 
-		public static void UpdateBarTextColor(this MauiToolbar nativeToolbar, Toolbar toolbar)
+		public static void UpdateBarTextColor(this MauiToolbar platformToolbar, Toolbar toolbar)
 		{
 			if (toolbar.BarTextColor != null)
-				nativeToolbar.TitleColor = toolbar.BarTextColor.ToPlatform();
+				platformToolbar.TitleColor = toolbar.BarTextColor.ToPlatform();
 		}
 
-		public static void UpdateToolbarDynamicOverflowEnabled(this MauiToolbar nativeToolbar, Toolbar toolbar)
+		public static void UpdateToolbarDynamicOverflowEnabled(this MauiToolbar platformToolbar, Toolbar toolbar)
 		{
-			if (nativeToolbar.CommandBar == null)
+			if (platformToolbar.CommandBar == null)
 				return;
 
-			nativeToolbar.CommandBar.IsDynamicOverflowEnabled = toolbar.DynamicOverflowEnabled;
+			platformToolbar.CommandBar.IsDynamicOverflowEnabled = toolbar.DynamicOverflowEnabled;
 		}
 	}
 }

--- a/src/Controls/src/Core/ShellToolbar.cs
+++ b/src/Controls/src/Core/ShellToolbar.cs
@@ -94,10 +94,6 @@ namespace Microsoft.Maui.Controls
 
 			UpdateTitle();
 
-			TitleView = _shell.GetEffectiveValue<VisualElement>(
-				Shell.TitleViewProperty,
-				Shell.GetTitleView(_shell));
-
 			if (_currentPage != null &&
 				_currentPage.IsSet(Shell.NavBarIsVisibleProperty))
 			{
@@ -169,8 +165,17 @@ namespace Microsoft.Maui.Controls
 
 		internal void UpdateTitle()
 		{
-			Page currentPage = _shell.GetCurrentShellPage() as Page;
+			TitleView = _shell.GetEffectiveValue<VisualElement>(
+				Shell.TitleViewProperty,
+				Shell.GetTitleView(_shell));
 
+			if (TitleView != null)
+			{
+				Title = String.Empty;
+				return;
+			}
+
+			Page currentPage = _shell.GetCurrentShellPage() as Page;
 			if (currentPage?.IsSet(Page.TitleProperty) == true)
 			{
 				Title = currentPage.Title ?? String.Empty;

--- a/src/Controls/src/Core/Toolbar.cs
+++ b/src/Controls/src/Core/Toolbar.cs
@@ -23,6 +23,8 @@ namespace Microsoft.Maui.Controls
 		bool _backButtonVisible;
 		bool _backButtonEnabled = true;
 		bool _drawerToggleVisible;
+		Maui.IElement _parent;
+		IElementHandler _handler;
 
 		public Toolbar(Maui.IElement parent)
 		{
@@ -43,10 +45,20 @@ namespace Microsoft.Maui.Controls
 		public bool BackButtonEnabled { get => _backButtonEnabled; set => SetProperty(ref _backButtonEnabled, value); }
 		public virtual bool DrawerToggleVisible { get => _drawerToggleVisible; set => SetProperty(ref _drawerToggleVisible, value); }
 		public bool IsVisible { get => _isVisible; set => SetProperty(ref _isVisible, value); }
-		public IElementHandler Handler { get; set; }
+		public IElementHandler Handler
+		{
+			get => _handler;
+			set
+			{
+				if (_handler == value)
+					return;
 
-		Maui.IElement _parent;
+				OnHandlerChanging(_handler, value);
+				_handler = value;
+			}
+		}
 
+		partial void OnHandlerChanging(IElementHandler oldHandler, IElementHandler newHandler);
 		public event PropertyChangedEventHandler PropertyChanged;
 
 		public Maui.IElement Parent => _parent;

--- a/src/Controls/tests/Core.UnitTests/ShellToolbarTests.cs
+++ b/src/Controls/tests/Core.UnitTests/ShellToolbarTests.cs
@@ -188,5 +188,26 @@ namespace Microsoft.Maui.Controls.Core.UnitTests
 
 			Assert.AreEqual(page.BindingContext, backButtonBehavior.BindingContext);
 		}
+
+		[Test]
+		public async Task TitleAndTitleViewAreMutuallyExclusive()
+		{
+			var contentPage = new ContentPage() { Title = "Test Title" };
+			var titleView = new VerticalStackLayout();
+
+			TestShell testShell = new TestShell(contentPage);
+			var window = new Window()
+			{
+				Page = testShell
+			};
+
+			var toolbar = testShell.Toolbar;
+			Assert.AreEqual("Test Title", toolbar.Title);
+			Shell.SetTitleView(contentPage, titleView);
+			Assert.IsEmpty(toolbar.Title);
+			Assert.AreEqual(titleView, toolbar.TitleView);
+			Shell.SetTitleView(contentPage, null);
+			Assert.AreEqual("Test Title", toolbar.Title);
+		}
 	}
 }

--- a/src/Controls/tests/Core.UnitTests/ToolbarTests.cs
+++ b/src/Controls/tests/Core.UnitTests/ToolbarTests.cs
@@ -38,6 +38,25 @@ namespace Microsoft.Maui.Controls.Core.UnitTests
 		}
 
 		[Test]
+		public async Task TitleAndTitleViewAreMutuallyExclusive()
+		{
+			var window = new Window();
+			IToolbarElement toolbarElement = window;
+			var contentPage = new ContentPage() { Title = "Test Title" };
+			var navigationPage = new NavigationPage(contentPage);
+			window.Page = navigationPage;
+
+			var titleView = new VerticalStackLayout();
+			var toolbar = (Toolbar)toolbarElement.Toolbar;
+			Assert.AreEqual("Test Title", toolbar.Title);
+			NavigationPage.SetTitleView(contentPage, titleView);
+			Assert.IsEmpty(toolbar.Title);
+			Assert.AreEqual(titleView, toolbar.TitleView);
+			NavigationPage.SetTitleView(contentPage, null);
+			Assert.AreEqual("Test Title", toolbar.Title);
+		}
+
+		[Test]
 		public async Task InsertPageBeforeRootPageShowsBackButton()
 		{
 			var window = new Window();

--- a/src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.cs
+++ b/src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.cs
@@ -122,6 +122,7 @@ namespace Microsoft.Maui.DeviceTests
 
 			var page1 = new ContentPage();
 			var page2 = new ContentPage();
+			var page3 = new ContentPage();
 
 			var titleView1 = new VerticalStackLayout();
 			var titleView2 = new Label();
@@ -150,7 +151,7 @@ namespace Microsoft.Maui.DeviceTests
 						new ShellContent()
 						{
 							Route = "Item3",
-							Content = new ContentPage()
+							Content = page3
 						},
 					}
 				});
@@ -158,6 +159,7 @@ namespace Microsoft.Maui.DeviceTests
 
 			await CreateHandlerAndAddToWindow<ShellHandler>(shell, async (handler) =>
 			{
+				await OnLoadedAsync(page1);
 				// GotoAsync which switching tabs/flyout items currently
 				// doesn't resolve after navigated has finished which is why we have the
 				// delays
@@ -165,15 +167,19 @@ namespace Microsoft.Maui.DeviceTests
 				Assert.Equal(titleView1.ToPlatform(), GetTitleView(handler));
 				await shell.GoToAsync("//Item2");
 				await Task.Delay(200);
+				await OnLoadedAsync(page2);
 				Assert.Equal(titleView2.ToPlatform(), GetTitleView(handler));
 				await shell.GoToAsync("//Item1");
 				await Task.Delay(200);
+				await OnLoadedAsync(page1);
 				Assert.Equal(titleView1.ToPlatform(), GetTitleView(handler));
 				await shell.GoToAsync("//Item2");
 				await Task.Delay(200);
+				await OnLoadedAsync(page2);
 				Assert.Equal(titleView2.ToPlatform(), GetTitleView(handler));
 				await shell.GoToAsync("//Item3");
 				await Task.Delay(200);
+				await OnLoadedAsync(page3);
 				Assert.Equal(shellTitleView.ToPlatform(), GetTitleView(handler));
 			});
 		}

--- a/src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.cs
+++ b/src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.cs
@@ -114,6 +114,70 @@ namespace Microsoft.Maui.DeviceTests
 			});
 		}
 
+
+		[Fact(DisplayName = "TitleView Causes a Crash When Switching Tabs")]
+		public async Task TitleViewCausesACrashWhenSwitchingTabs()
+		{
+			SetupBuilder();
+
+			var page1 = new ContentPage();
+			var page2 = new ContentPage();
+
+			var titleView1 = new VerticalStackLayout();
+			var titleView2 = new Label();
+			var shellTitleView = new Editor();
+
+			Shell.SetTitleView(page1, titleView1);
+			Shell.SetTitleView(page2, titleView2);
+
+			var shell = await CreateShellAsync((shell) =>
+			{
+				Shell.SetTitleView(shell, shellTitleView);
+				shell.Items.Add(new TabBar()
+				{
+					Items =
+					{
+						new ShellContent()
+						{
+							Route = "Item1",
+							Content = page1
+						},
+						new ShellContent()
+						{
+							Route = "Item2",
+							Content = page2
+						},
+						new ShellContent()
+						{
+							Route = "Item3",
+							Content = new ContentPage()
+						},
+					}
+				});
+			});
+
+			await CreateHandlerAndAddToWindow<ShellHandler>(shell, async (handler) =>
+			{
+				// GotoAsync which switching tabs/flyout items currently
+				// doesn't resolve after navigated has finished which is why we have the
+				// delays
+				// https://github.com/dotnet/maui/issues/6193
+				Assert.Equal(titleView1.ToPlatform(), GetTitleView(handler));
+				await shell.GoToAsync("//Item2");
+				await Task.Delay(200);
+				Assert.Equal(titleView2.ToPlatform(), GetTitleView(handler));
+				await shell.GoToAsync("//Item1");
+				await Task.Delay(200);
+				Assert.Equal(titleView1.ToPlatform(), GetTitleView(handler));
+				await shell.GoToAsync("//Item2");
+				await Task.Delay(200);
+				Assert.Equal(titleView2.ToPlatform(), GetTitleView(handler));
+				await shell.GoToAsync("//Item3");
+				await Task.Delay(200);
+				Assert.Equal(shellTitleView.ToPlatform(), GetTitleView(handler));
+			});
+		}
+
 		[Fact(DisplayName = "Handlers not recreated when changing tabs")]
 		public async Task HandlersNotRecreatedWhenChangingTabs()
 		{

--- a/src/Controls/tests/DeviceTests/HandlerTestBase.Android.cs
+++ b/src/Controls/tests/DeviceTests/HandlerTestBase.Android.cs
@@ -111,7 +111,7 @@ namespace Microsoft.Maui.DeviceTests
 		protected AView GetTitleView(IElementHandler handler)
 		{
 			var toolbar = GetPlatformToolbar(handler);
-			var container = toolbar.GetFirstChildOfType<Controls.Toolbar.Container>();
+			var container = toolbar?.GetFirstChildOfType<Controls.Toolbar.Container>();
 
 			if (container != null && container.ChildCount > 0)
 				return container.GetChildAt(0);

--- a/src/Controls/tests/DeviceTests/HandlerTestBase.Android.cs
+++ b/src/Controls/tests/DeviceTests/HandlerTestBase.Android.cs
@@ -108,6 +108,17 @@ namespace Microsoft.Maui.DeviceTests
 			});
 		}
 
+		protected AView GetTitleView(IElementHandler handler)
+		{
+			var toolbar = GetPlatformToolbar(handler);
+			var container = toolbar.GetFirstChildOfType<Controls.Toolbar.Container>();
+
+			if (container != null && container.ChildCount > 0)
+				return container.GetChildAt(0);
+
+			return null;
+		}
+
 		protected MaterialToolbar GetPlatformToolbar(IElementHandler handler)
 		{
 			if (handler is Microsoft.Maui.Controls.Handlers.Compatibility.ShellRenderer sr)

--- a/src/Controls/tests/DeviceTests/HandlerTestBase.Windows.cs
+++ b/src/Controls/tests/DeviceTests/HandlerTestBase.Windows.cs
@@ -145,5 +145,11 @@ namespace Microsoft.Maui.DeviceTests
 			MauiToolbar windowHeader = (MauiToolbar)navView.Header;
 			return windowHeader;
 		}
+
+		protected object GetTitleView(IElementHandler handler)
+		{
+			var toolbar = GetPlatformToolbar(handler);
+			return toolbar.TitleView;
+		}
 	}
 }

--- a/src/Controls/tests/DeviceTests/HandlerTestBase.iOS.cs
+++ b/src/Controls/tests/DeviceTests/HandlerTestBase.iOS.cs
@@ -78,5 +78,10 @@ namespace Microsoft.Maui.DeviceTests
 
 			throw new NotImplementedException();
 		}
+
+		protected object GetTitleView(IElementHandler handler)
+		{
+			throw new NotImplementedException();
+		}
 	}
 }

--- a/src/Core/src/Platform/Windows/MauiToolbar.xaml.cs
+++ b/src/Core/src/Platform/Windows/MauiToolbar.xaml.cs
@@ -60,6 +60,12 @@ namespace Microsoft.Maui.Platform
 			}
 		}
 
+		internal UI.Xaml.Thickness TitleViewMargin
+		{
+			get => titleView.Margin;
+			set => titleView.Margin = value;
+		}
+
 		internal object? TitleView
 		{
 			get => titleView.Content;

